clang/test/CodeGenSYCL/sycl-device-const-sdm.cpp

```diff
@@ -0,0 +1,91 @@
+// RUN: %clang_cc1 -fsycl -fsycl-is-device -triple spir64-unknown-unknown-sycldevice -disable-llvm-passes %s -emit-llvm -o -  | FileCheck %s
+// Tests that static data members that are not constant-initialized are not
+// emitted in device code.
+
+// CHECK:%struct._ZTS1B.B = type { i8 }
+// CHECK:%struct._ZTS1C.C = type { i32, i32 }
+// CHECK:%struct._ZTS1D.D = type { i8 }
+// CHECK:$_ZN2TTIiE2c1E = comdat any
+// CHECK:$_ZN2TTIiE2d1E = comdat any
+// CHECK:$_ZN2TTIiE4var1E = comdat any
+// CHECK:$_ZN2TTIiE4var3E = comdat any
+// CHECK:@_ZN1S2b1E = addrspace(1) global %struct._ZTS1B.B zeroinitializer, align 1
+// CHECK:@_ZN1S2c1E = addrspace(1) constant %struct._ZTS1C.C { i32 2, i32 5 }, align 4
+// CHECK:@_ZN1S2d1E = addrspace(1) constant %struct._ZTS1D.D zeroinitializer, align 1
+// CHECK:@_ZN1S4var1E = addrspace(1) constant i32 1, align 4
+// CHECK:@_ZN1S4var2E = available_externally addrspace(1) constant i32 3, align 4
+// CHECK:@_ZN1S4var3E = addrspace(1) constant i32 4, align 4
+// CHECK:@_ZN2TTIiE2b1E = external addrspace(1) global %struct._ZTS1B.B, align 1
+// CHECK:@_ZN2TTIiE2c1E = linkonce_odr addrspace(1) constant %struct._ZTS1C.C { i32 2, i32 5 }, comdat, align 4
+// CHECK:@_ZN2TTIiE2d1E = linkonce_odr addrspace(1) constant %struct._ZTS1D.D zeroinitializer, comdat, align 1
+// CHECK:@_ZN2TTIiE4var1E = linkonce_odr addrspace(1) constant i32 1, comdat, align 4
+// CHECK:@_ZN2TTIiE4var2E = available_externally addrspace(1) constant i32 3, align 4
+// CHECK:@_ZN2TTIiE4var3E = linkonce_odr addrspace(1) constant i32 4, comdat, align 4
+// CHECK:@llvm.global_ctors = appending global [1 x { i32, void ()*, i8* }] [{ i32, void ()*, i8* } { i32 65535
```

Comment:
It is still needed?

Reply:
Not strictly needed.   This one is the call to the top-level routine that in turn calls __cxx_global_var_init*, which in turn calls the global constructors.   I have suppressed the last, so this becomes a no-op.   I will see if I can remove it entirely.

Reply:
I have added diagnostics for the use of const static data members that are used in the device.   With that, we no longer emit the initializers for these, and the llvm.global_ctors mechanism is not generated.